Review PR

ottomata opened the pull request.

ottomata committed:
Updates

ottomata committed:
Add tests

ottomata committed:
Fix deps

ottomata committed:
Refactor - no longer use git to find schema versions, use schema version field instead

ottomata committed:
Linting

ottomata committed:
Removed unused .gitmodules

ottomata committed:
Remove uneeded objectFactory

ottomata committed:
Slight refactor to arg vs stdin schema reading

ottomata committed:
Run git add command if files changed

This works better with a pre commit hook than a gitattributes filter

ottomata committed:
Remove extra newlines

ottomata committed:
more

ottomata committed:
WIP

ottomata committed:
Added CLI for materializing modified files

ottomata committed:
Support generating both json and yaml

ottomata committed:
Update README

ottomata committed:
README updates

ottomata committed:
README updates

ottomata committed:
Add $ref dereference support

ottomata committed:
Add tests for materializing dereferenced schemas

ottomata committed:
Don't re-load schema fixture from disk every test

ottomata committed:
Remove comment

ottomata committed:
Doc updates

Pchelolo commented on .eslintrc.yml:
```diff
@@ -0,0 +1,33 @@
+  quote-props: off
+  operator-linebreak: off
+  max-statements-per-line: off
+  brace-style: off
```
Hm... that's quite a lot of overrides. Why wouldn't we closer follow WMF style guide?

ottomata replied:
I copy/pasted this from somewhere...no reason I know of. will remove whatever you think is needed!

ottomata replied:
Removed some of these.

Pchelolo commented on README.md:
````diff
@@ -1,8 +1,90 @@
+
+# Usage
+```
+$ npm install jsonschema-tools
````
Perhaps in the example we could suggest installing it globally with `-g` option. Then we wouldn't need to add it to path.

ottomata replied:
Done

Pchelolo commented on test/test.js:
```diff
@@ -0,0 +1,196 @@
+        type: 'string',
+        format: 'date-time',
+    },
+};
```
Can we load it up from the real common/current.yaml instead of duplicating?

ottomata replied:
Done

Pchelolo commented on scripts/jsonschema-tools.js:
```diff
@@ -0,0 +1,333 @@
+async function dereference(args) {
+    const options = argsToOptions(args);
+
+    if (_.isEmpty(args.schemaPath)) {
```
Since we've converted to options, let's use them now for clarity. Now it's confusing why we're still using args.

ottomata replied:
Done.

Pchelolo commented on scripts/jsonschema-tools.js:
```diff
@@ -0,0 +1,333 @@
+    const options = argsToOptions(args);
+
+    if (_.isEmpty(args.schemaPath)) {
+        args.schemaPath.push(0);
```
Emmm??

ottomata replied:
Read from stdin.  Adding comment.

Pchelolo commented on README.md:
```diff
@@ -1,8 +1,90 @@
+  [schema-path...]                          versioned files.
+  jsonschema-tools materialize-modified     Looks for git modified JSONSchema
+  [git-root]                                files and materializes them.
+  jsonschema-tools install-git-hook         Installs a git pre-commit hook that
```
There definitely should be `remove-git-hook` too.

ottomata replied:
Good point! ok!

ottomata replied:
Hm, on second thought, remove-git-hook is nothing more than `rm .git/hooks/pre-commit`.  Does the tool need to do this?

Pchelolo reviewed (commented):
A couple of minor comments. Will play with it more l8r

ottomata committed:
Review fixes for PR #1

ottomata committed:
More PR #1 fixes

ottomata committed:
Fix schemaPath arg -> option

ottomata committed:
Fix unnecessary quotes

ottomata committed:
Remove unnecessary dependency

ottomata committed:
package.json updates

ottomata committed:
Add LICENSE

ottomata committed:
Change npm package name to scoped @wikimedia/jsonschema-tools

ottomata committed:
Properly require fs-extra and pino-pretty in dependencies

ottomata committed:
Properly specify which files to publish in package.json

ottomata committed:
Include .eslintrc.yml in publish files

ottomata committed:
Fix missing comma

ottomata committed:
Properly require @wikimedia/jsonschema-tools in git pre-commit hook

ottomata committed:
Use indendeted json serializer

ottomata committed:
Fail kindly in pre-commit hook if jsonschema-tools is not installed

Bump version to 0.0.7

ottomata committed:
README updates

ottomata committed:
Materialize modified schemas in serial

This helps avoid attempting to materialize a schema before its $ref dependency
is materialized.

Also add better error messages.

Bump to 0.1.0.

ottomata committed:
Merge pull request #2 from ottomata/otto

Materialize modified schemas in serial

ottomata committed:
Set default log level for git commit to info

ottomata committed:
Use json-schema-merge-allof to merge allOf fields

This allows for schemas to more easily include each other using
$ref and allOf.  E.g.

allOf:
  - $ref: /path/to/common/schema/1.0.0
  - properties:
      myfieldA:
        type: string
    ...

Dereferencing this schema will now also  merge everything from the common schema
with the properties for this schema.

ottomata committed:
Fix variable casing

ottomata committed:
Modify tests to dereference and merge allOf

ottomata committed:
Remove console.log

ottomata committed:
Merge pull request #3 from ottomata/otto3

Use json-schema-merge-allof to merge allOf fields

ottomata committed:
Readme updates

ottomata committed:
Merge branch 'master' of github.com:wikimedia/jsonschema-tools

ottomata committed:
Add functions to find schemas, examine them, and group them

This will help with creating tests to ensure that schemas materialize
into what is expected, and to potentially help create a 'materialize-all'
type functionality.

Bug: T206814

ottomata committed:
Use consistent variable name for schemaInfosByTitle

ottomata committed:
Remove unneeded dependency on 'file'

ottomata committed:
Merge pull request #4 from ottomata/master

Add functions to find schemas, examine them, and group them

ottomata committed:
Add materialize-all command

This also makes better guesses at how to sort schema files,
hopefully materializing dependency ones earlier in the loop.

Commit:
Make schema discovery functions syncronous.

ottomata committed:
Merge pull request #6 from Pchelolo/sync

Add syncronous versions of the schema discovery funtions

ottomata committed:
Use path hierarchy depth as sort critera instead of title compare

This is consistent with the sorting used by materializeModifiedSchemas,
where title is not available.

Commit:
Merge pull request #5 from ottomata/master

Add materialize-all command

Commit:
Expose findSchemasByTitle

ottomata committed:
Merge pull request #7 from Pchelolo/more_discovery

Expose findSchemasByTitle

ottomata committed:
Support loading options from config files, refactor git related commands

In order to ease use of jsonschema-tools as a library, this commit
adds support for reading in configs from config files, e.g.
.jsonschema-tools.yaml.  This yaml or json config file is an
object keyed by the same options described in the defaultOptions
documentation in index.js.

defaultOptions are always set, and then overridden by
provided config files, and then lastly by any CLI options
if running in CLI mode.  The config file will allow
easier configuration of schema repositories, without having
to manually provide CLI opts when running on the CLI (or
from the pre-commit hook), or when running as a JS library,
e.g. for tests.

Some CLI opts and function args have changed.

- git-root is no longer a CLI opt. When needed, git root
  will be discovered from the schema-base-path.

- The pre-commit hook installed by install-git-hook no longer
  renders custom options from the CLI as javascript.  If
  you need to customize the pre-commit hook, you should
  edit the .jsonschema-tools.yaml (or other) config file.

- materialize-modified will now by default look for unstaged
  schema files.  Staged ones likely only should be materialized
  during the pre-commit hook, so the pre-commit hook
  overrides this by passing --staged.

- materialize-modified takes an optional schema-base-path
  instead of the path to the git-root.

Note: the new readConfig function by default will only attempt
to read config files the first time it is called.  After that,
it expects that options has already been set by a previous
call to readConfig.  This allows to pass options around without
keeping global state by reading config the first time it is needed.
If you need to re-read config, just set the force arg to true.

ottomata committed:
Merge pull request #8 from ottomata/master

Support loading options from config files; refactor git related commands

ottomata committed:
Remove all schemaBasePath params; this has been moved into options

ottomata committed:
Move code to lib/jsonschema-tools.js; use index.js for explicit module exports

ottomata committed:
Fix test to provide schemaBasePath in options

ottomata committed:
[WIP] Add lib/tests.js to make repository tests exportable

Run these tests on the included fixtures here.

ottomata committed:
Adjust schema structure tests to use schemaInfo path dir

ottomata committed:
Move repository test libraries into lib/tests and export as .tests.*

ottomata committed:
Add schema version compatibility tests

ottomata committed:
Add robustness (aka convention) tests

ottomata committed:
Update README for repo tests and move tests module exports to lib/tests/index.js

ottomata committed:
Bump to 0.4.0

ottomata committed:
Merge pull request #1 from ottomata/module-tests

Exportable Schema Repository Tests

ottomata committed:
Merge pull request #9 from ottomata/master

Exportable schema repository tests

ottomata committed:
Don't skip monomorphic tests

ottomata committed:
Fix to include lib/ dir in npm package

Bump to 0.4.1

ottomata committed:
Require ajv

Release 0.4.2

ottomata committed:
Add missing test config files

ottomata committed:
Fix test expected missing maxLength

ottomata committed:
Reduce pino-pretty log noise

ottomata committed:
Add ignoreSchemas option, use to ignore schemas with certain $id

This will allow for manually configuring tests to skip
certain schema versions that do not conform to the
test requirements.

Bump to 0.5.0

ottomata committed:
Merge pull request #10 from ottomata/ignore

Add ignoreSchemas option, use to ignore schemas with certain $id

ottomata committed:
Add getSchemaById function to use schemaBaseUrls to return schema by $id

Bump to version 0.6.0

ottomata committed:
Merge pull request #11 from ottomata/master

Add getSchemaById function to use schemaBaseUrls to return schema by $id

ottomata committed:
Create 'latest' symlink(s) when materializing schemas

Add option that if true, create's both latest.(json|yaml) as well as an
extensionless 'latest' symlink (if configured to do so) when materializing
schemas.  This symlinks will only ever be updated if jsonschema-tools
matereializes a schema that has a greater version than the version
of the schema that the latest symlink currently points at.

Bug: T245859

ottomata committed:
Fix some comment wording

ottomata committed:
Fix wording in README

ottomata committed:
Merge pull request #12 from ottomata/master

Create 'latest' symlink(s) when materializing schemas

ottomata committed:
Fix --no-symlink* args and clarify use of --no-* args with yargs in argsToOptions

ottomata committed:
Create proper relativee latest symlinks

ottomata committed:
In test library, check that JSONSchema examples validate

Bug: T242454

ottomata committed:
Also check that examples $schema matches schema $id

Bug: T242454

ottomata committed:
Don't reinstantiate ajv in test loop, just remove cached schema

Bug: T242454

ottomata committed:
README fixes

ottomata committed:
README fix for yaml formatting

ottomata committed:
Merge pull request #13 from ottomata/master

Validate examples against their schema

ottomata committed:
Make sure latestSchemaInfo is correctly chosen in repository structure test

ottomata committed:
Fix materialize CLI schema paths parsing

ottomata committed:
Structure repo test should ensure to compare latest version with current version $id

Fixes a bug where the chosen latest schema version was not correct.

ottomata committed:
Sort yaml schema keys when serializing

Bump to version 0.7.6

ottomata committed:
Fixes for structure tests when schema title does not match relative directory

The schema repository tests are dynamic.  The structure tests are grouped
by schema_title.  If something has gone wrong, and there is a materialized
schema file that has a title that does not match the directory it is in,
furhter test declarations will fail.

Prevent further test declarations if any schema titles and directories
are mismatched.

Bump to 0.7.7.

Commit:
Add option to enforce bounds for number fields

This option configures a pair of numeric bounds that, at the
schema materialization level, will apply those bounds to number
fields (float, integer or number) by traversing the schema.

The change includes a small refactor that separates the logic that
modifies the materialized schema from the main function that
outputs the materialized schema to a path.

Bug: T258659

Commit:
Apply CR comments

Commit:
Make tests fail instead of forcing new bounds

Commit:
Correct typo and add test to check existence of max and min in materialized schemas

ottomata committed:
Merge pull request #16 from wikimedia/number-limits

Add option to enforce bounds for number fields

cdanis committed:
properly parse versions from $ids with numbers

Previously, when $id strings were passed directly to semver.coerce(),
$ids like '/w3c/reportingapi/report/1.2.3' would have been parsed as
version number '3.0.0'.

ottomata committed:
Use basename in schemaVersion rather than splitting path

ottomata committed:
Fix bug in enforced numeric bounds if max or min is 0

- Fix bug where a min or max of 0 was being interpreted as false and overwritten with bound
- Fix tests to handle this

Bug: T258659

ottomata committed:
Merge pull request #18 from ottomata/enforcedbounds

Fix bug in enforced numeric bounds if max or min is 0

ottomata committed:
Merge pull request #17 from cdanis/semver-embedded-numbers

properly parse versions from $ids with numbers

ottomata committed:
Add .github workflow for npm test CI

ottomata commented:
This PR was just used for initial commit review.  Closing.
